### PR TITLE
CLI: convert sharded PT models

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR adds a major upgrade and a minor change to the `pt-to-tf` CLI.

Major upgrade: we can now convert sharded PT models. It updates how the `from_pt` loading works so as to be able to load from shards. It also updates how the `pt-to-tf` CLI stores the models, so it uses sharding capabilities when needed.

Minor change: adds a flag to control the maximum hidden layer admissible error. It is relatively common to find models where the outputs from the PT and TF models are nearly the same, but the hidden layers have a larger mismatch. This flag allows us to temporarily increase the admissible error if the model seems to be behaving properly (for instance, all RegNet models had a hidden layer difference between 1e-4 and 1e-2, but the outputs were behaving properly).

Example of sharded TF model PR, using the updated tools: https://huggingface.co/facebook/regnet-y-10b-seer-in1k/discussions/1

### CODE REVIEW

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 1]
```diff
@@ -117,10 +117,17 @@ def load_pytorch_checkpoint_in_tf2_model(
...
+    # Treats a single file as a collection of shards with 1 shard.
+    if isinstance(pytorch_checkpoint_path, str):
+        pytorch_checkpoint_path = [pytorch_checkpoint_path]
+
+    # Loads all shards into a single state dictionary
+    pt_state_dict = {}
+    for path in pytorch_checkpoint_path:
+        pt_path = os.path.abspath(path)
+        logger.info(f"Loading PyTorch weights from {pt_path}")
+        pt_state_dict.update(torch.load(pt_path, map_location="cpu"))
```
Comment (ArthurZucker) [inline src/transformers/modeling_tf_pytorch_utils.py:129]: That's super nice 👍🏻

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for working on that!

**Review by Rocketknight1 — APPROVED** [review]
LGTM to me too!

**Comment by ArthurZucker** [comment]
BTW could we add 2 tests, `test_load_sharded_tf_to_pt` and `load_sharded_pt_to_tf`

File: src/transformers/modeling_tf_pytorch_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_tf_pytorch_utils.py:129]: That is a nice first step, but ideally, we'd want to convert the shards one by one to avoid using too much RAM and be able to convert LLMs checkpoints without needing a battle station.

**Comment by gante** [comment]
TF shards -> PT probably won't work, but I will add the test for PT shards -> TF 👍

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -2157,11 +2158,15 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
                 if from_pt and os.path.isfile(os.path.join(pretrained_model_name_or_path, WEIGHTS_NAME)):
                     # Load from a PyTorch checkpoint in priority if from_pt
                     archive_file = os.path.join(pretrained_model_name_or_path, WEIGHTS_NAME)
+                elif from_pt and os.path.isfile(os.path.join(pretrained_model_name_or_path, WEIGHTS_INDEX_NAME)):
+                    # Load from a sharded PyTorch checkpoint
+                    archive_file = os.path.join(pretrained_model_name_or_path, WEIGHTS_INDEX_NAME)
+                    is_sharded = True
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:2164]: Nice addition, maybe we should also support loading from a remote sharded checkpoint with `from_pt=True`? (It should be its own PR if we decide to support this.)

**Review by sgugger — APPROVED** [review]
Nice improvements, thanks!

File: src/transformers/modeling_tf_pytorch_utils.py (hunk 1 above)
Comment (gante) [reply src/transformers/modeling_tf_pytorch_utils.py:129]: haha yes, I had to spin up a machine with >100GB or RAM to convert the RegNets 😬
